packages/react-form/src/hooks/field/reducer.ts

```diff
@@ -55,62 +55,77 @@ export type FieldAction<Value> =
   | UpdateAction<Value>
   | NewDefaultAction<Value>;
 
-export function reduceField<Value>(
-  state: FieldState<Value>,
-  action: FieldAction<Value>,
-) {
-  switch (action.type) {
-    case 'update': {
-      const newValue = action.payload;
-      const {defaultValue} = state;
-      const isDirty = Array.isArray(defaultValue)
-        ? !shallowArrayComparison(defaultValue, newValue)
-        : defaultValue !== newValue;
-
-      return {
-        ...state,
-        dirty: isDirty,
-        value: newValue,
-        touched: true,
-      };
+export function makeFieldReducer<Value>({
+  dirtyStateComparator,
+}: {
+  dirtyStateComparator: DirtyStateComparator<Value>;
+}): Reducer<FieldState<Value>, FieldAction<Value>> {
+  return (state: FieldState<Value>, action: FieldAction<Value>) => {
+    switch (action.type) {
+      case 'update': {
+        const newValue = action.payload;
+        const {defaultValue} = state;
+        const dirty = dirtyStateComparator(defaultValue, newValue);
+
+        return {
+          ...state,
+          dirty,
+          value: newValue,
+          touched: true,
+        };
+      }
+
+      case 'updateError': {
+        return {
+          ...state,
+          error: action.payload,
+        };
+      }
+
+      case 'reset': {
+        const {defaultValue} = state;
+
+        return {
+          ...state,
+          error: undefined,
+          value: defaultValue,
+          dirty: false,
+          touched: false,
+        };
+      }
+
+      case 'newDefaultValue': {
+        const newDefaultValue = action.payload;
+        return {
+          ...state,
+          error: undefined,
+          value: newDefaultValue,
+          defaultValue: newDefaultValue,
+          touched: false,
+          dirty: false,
+        };
+      }
     }
+  };
+}
 
-    case 'updateError': {
-      return {
-        ...state,
-        error: action.payload,
-      };
-    }
-
-    case 'reset': {
-      const {defaultValue} = state;
-
-      return {
-        ...state,
-        error: undefined,
-        value: defaultValue,
-        dirty: false,
-        touched: false,
-      };
-    }
+const shallowFieldReducer = makeFieldReducer({
+  dirtyStateComparator: defaultDirtyComparator,
+});
 
-    case 'newDefaultValue': {
-      const newDefaultValue = action.payload;
-      return {
-        ...state,
-        error: undefined,
-        value: newDefaultValue,
-        defaultValue: newDefaultValue,
-        touched: false,
-        dirty: false,
-      };
-    }
-  }
+export function reduceField<Value>(
```

Comment:
this and the `shallowFieldReducer` feel like they may belong somewhere else?

Reply:
It feels like that to me too, but I just tried moving it out of that file and couldn't find a place that felt right because the call to `makeFieldReducer` would have to call back here anyway. 😕